src/hotspot/share/cds/heapShared.cpp

```diff
@@ -601,6 +641,104 @@ void HeapShared::set_has_native_pointers(oop src_obj) {
   info->set_has_native_pointers();
 }
 
+void HeapShared::start_finding_archivable_hidden_classes() {
+  if (!CDSConfig::is_dumping_invokedynamic()) {
+    return;
+  }
+  NoSafepointVerifier nsv;
+
+  init_seen_objects_table();
+
+  find_archivable_hidden_classes_helper(archive_subgraph_entry_fields);
+  if (CDSConfig::is_dumping_full_module_graph()) {
+    find_archivable_hidden_classes_helper(fmg_archive_subgraph_entry_fields);
+  }
+}
+
+void HeapShared::end_finding_archivable_hidden_classes() {
+  if (!CDSConfig::is_dumping_invokedynamic()) {
+    return;
+  }
+  NoSafepointVerifier nsv;
+
+  delete_seen_objects_table();
+}
+
+void HeapShared::find_archivable_hidden_classes_helper(ArchivableStaticFieldInfo fields[]) {
+  if (!CDSConfig::is_dumping_heap()) {
+    return;
+  }
+  for (int i = 0; fields[i].valid(); ) {
+    ArchivableStaticFieldInfo* info = &fields[i];
+    const char* klass_name = info->klass_name;
+    for (; fields[i].valid(); i++) {
+      ArchivableStaticFieldInfo* f = &fields[i];
```

Comment:
It looks like the pattern of nested for-loops is copied from HeapShared::archive_object_subgraphs()
but this pattern doesn't seem to give any advantage here. If so can we replace it with just single for loop to make it easier to understand?